client/diff.go

```diff
@@ -0,0 +1,273 @@
+
+import (
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"net/http"
+	"reflect"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/theupdateframework/notary"
+	store "github.com/theupdateframework/notary/storage"
+	"github.com/theupdateframework/notary/trustpinning"
+	"github.com/theupdateframework/notary/tuf"
+	"github.com/theupdateframework/notary/tuf/data"
+	"github.com/theupdateframework/notary/tuf/utils"
+)
+
+// Diff represents the difference between two versions of the same repo
+type Diff struct {
+	TargetsAdded   []TargetSignedStruct
+	TargetsRemoved []TargetSignedStruct
+	TargetsUpdated []TargetSignedStruct
+	RolesAdded     []data.Role
+	RolesRemoved   []data.Role
+	RolesUpdated   []data.Role
+}
+
+// NewDiff returns the different between two versions of the same TUF repo
+func NewDiff(gun data.GUN, baseURL string, rt http.RoundTripper, hash1, hash2 string) (Diff, error) {
+	remoteStore, err := getRemoteStore(baseURL, gun, rt)
+	if err != nil {
+		// baseURL is syntactically invalid
+		return Diff{}, err
+	}
+
+	first, err := setupClient(gun, hash1, remoteStore)
+	if err != nil {
+		return Diff{}, err
+	}
+	second, err := setupClient(gun, hash2, remoteStore)
+	if err != nil {
+		return Diff{}, err
+	}
+
+	// download the relevant versions of the repos now that we have the bootstrapping
+	// setup.
+	// N.B. lower case update will fail out immediately rather than attempting
+	//      to download a newer root
+	logrus.Debug("starting updates")
+	err = first.update()
+	if err != nil {
+		return Diff{}, errors.New("failed to download repo version " + hash1)
+	}
+	logrus.Debug("succeeded updating first")
+	err = second.update()
+	if err != nil {
+		return Diff{}, errors.New("failed to download repo version " + hash2)
+	}
+	logrus.Debug("succeeded updated second")
+
+	repoFirst, _, err := first.newBuilder.Finish()
+	repoSecond, _, err := second.newBuilder.Finish()
+
+	return diff(repoFirst, repoSecond)
+}
+
+func setupClient(gun data.GUN, tsHash string, remote store.RemoteStore) (*tufClient, error) {
+	hashBytes, err := hex.DecodeString(tsHash)
+	if err != nil {
+		return nil, err
+	}
+
+	root, err := findRoot(hashBytes, remote)
+
+	oldBuilder := tuf.NewRepoBuilder(gun, nil, trustpinning.TrustPinConfig{})
+	if err := oldBuilder.Load(data.CanonicalRootRole, root, 0, true); err != nil {
+		return nil, err
+	}
+	newBuilder := tuf.NewRepoBuilder(gun, nil, trustpinning.TrustPinConfig{})
+	if err := newBuilder.Load(data.CanonicalRootRole, root, 0, true); err != nil {
+		return nil, err
+	}
+
+	tClient := newTufClient(oldBuilder, newBuilder, remote, store.NewMemoryStore(nil))
+	tClient.tsChecksum = hashBytes
+	return tClient, nil
+}
+
+func diff(first, second *tuf.Repo) (Diff, error) {
+	res := Diff{}
+	err := diffTargets(&res, first, second)
+	if err != nil {
+		return res, err
+	}
+	err = diffRoles(&res, first, second)
+	return res, err
+}
+
+func diffRoles(res *Diff, first, second *tuf.Repo) error {
+	firstRoles := first.GetAllLoadedRoles()
```

Comment:
Did you have a reason for using this function instead of `ListRoles`?  This should be fine given the ordering of operations (we have an `Update` when getting targets), though I'm wondering if we should add a comment to ensure that the code doesn't get shifted around

Reply:
I think ListRoles only returned delegations, or at least appeared to when I was tired last night

Reply:
ah - it should return all roles.  This code is fine as-is but `ListRoles` may be a safer improvement